### ISSUE TITLE
Create and enhance registration service type

### ISSUE THREAD

**Issue by uwe-salomon** [ARINC-IA/CSMIM#8]
**Request summary:**
Create an object type for the registration service as standardized by CSMIM.
Enhance the registration service by a lookup feature and add that to the type description.

**Security Domain:**
core

**Existing Types:**
None, but a description in the CSMIM standard.

**New Types:**
`csmim.obj.registration.1`
`csmim.obj.directory.1` or join with registration service.

**Features:**

Regarding the lookup feature: We already have the object ICDs being published in the network. This is sufficient if a particular object type of interest is only instantiated in few, well-known paths. If, on the other side, an object type can be instantiated in many places, a consumer would have to subscribe `#/icd/#` or similar, in consequence receiving many ICD messages and having to parse them to look for the object type of interest. Generic services like fault reporting or config reporting are particularly impacted by this problem. Every registration and re-registration in the network would send out a volley of ICD messages to these generic services.

One potential solution is to add a "directory service" to the network. This service would most likely be an intrinsic part of the registration service for efficiency reasons. Clients can send requests to the directory service, in order to be alerted when a new registration with certain properties is made. In particular, clients can request to be informed about all objects that have a certain object type of interest. This removes the need to subscribe to all ICD topics; a client would be informed only about the few registrations that are of concern actually.

As an example, a fault reporting service (a.k.a central maintenance service) could want to be informed about all objects that have the type csmim.obj.fault-report.1, wherever they are located in the network, but not from the pax domain.

This idea is similar to the CoRE Resource Directory [[RFC 9176](https://datatracker.ietf.org/doc/rfc9176/)]. We might benefit from looking at their design and experience.

**Comment by uwe-salomon** [comment]
As a side note, we also discussed other solutions to this problem. One idea was to use a specifically prefixed object path for these "generic" object types. For example, a galley object might be published under the path `crew/galley/galley/X1`. But its fault report object would be published under the path `crew/galley/faults/galley/X1`. This way, a fault reporting service could directly subscribe the topic `v1/data/crew/+/faults/#` to get all fault reports in the crew domain. Which would be superior to first receiving all ICDs and then having to subscribe the individual objects of interest.

A drawback of this solution is that it requires a special prefix (such as "faults" in the example above) for all such situations. This prefix must be agreed on in advance. The object tree must be kept clean below this prefix. And proper use of the CSMIM protocol would actually require each client to verify the matching object type before accessing the resources.

**Comment by matthias-heinisch** [comment]
I like the proposal and think it is very useful. Especially as we expect various of these pervasive topics (configuration, health status, security logging, ...). But I am not convinced this should be integrated into registration service, as handling of individual requests and distribution lists will be complex. Maybe we can define a filter set that we provide instead of these individual queries? For example the registration could publish all ICDs and the directory service compiles different pre-defined subsets.

**Comment by uwe-salomon** [comment]
A concrete proposal for the topic structure:

- The directory service registers its object under `core/csmim/directory`
- This object currently has one resource `csmim.obj`
- Under this resource, each object registration is published in its own topic using the CSMIM extended topic feature. The extended topic consists of the object type ID (minus the "csmim.obj" prefix, and splitting the version suffix with a slash instead of a dot) and the path of that object.
- The payload of each published topic could be a timestamp or the actual object type ID.

**Example**

Say that some device registered a new object under the path `crew/galley/left/x5` with the object type `csmim.obj.galley.1` and the supertype `csmim.obj.fault-report.1`. This supertype is configured in the directory service to be a type of interest. Then the directory service publishes the following message:

```
PUBLISH v1/data/core/csmim/directory/csmim.obj/fault-report/1/crew/galley/left/x5
QoS: 1
Retain: true
Payload:
  csmim.obj.galley.1
```

A fault reporting service would have to

- subscribe `v1/data/core/csmim/directory/csmim.obj/fault-report/#` in order to obtain a message for each registered object that derives from "csmim.obj.fault-report.N"
- parse the extended topic to retrieve the object type version number and the object path
- verify that it knows the "csmim.obj.fault-report.N" version
- subscribe to the announced faults resource

In the example, the fault reporting service would have to subscribe to the `crew/galley/left/x5/faults` resource (assuming that a fault reporter is defined to contain such a resource `faults`).

**Notes**

- The version number for object type IDs is optional. The directory service would have to detect a missing version number and insert an empty `//` into the topic.

**Comment by uwe-salomon** [comment]
Here are my results from testing a preliminary implementation.

![resource-directory-test](https://github.com/user-attachments/assets/26d593a2-238a-4a84-bada-5a1e33385cd8)

First, a description of the implementation itself. I implemented the CSMIM central services as a plugin for the open-source Mosquitto broker, written in C++. That means, the authentication service, registration service and the new resource directory service are a small shared library which is loaded when the broker starts. Consequently, they all run in the same address space, and use a shared data structure to authenticate and register CSMIM endpoints and do permissions checking. The implementation uses roughly 3 threads (1 broker, 1 registration service and resource directory, 1 authentication service). Synchronization is done via one central mutex. The resource directory was implemented as proposed in my previous comment. The resource directory looks only for objects derived from "csmim.obj.fault-report.1" and publishes their paths as described.

I implemented a test application that consists of two parts: a simulation of many CSMIM endpoints and a fault reporting service. Each simulated endpoint connects to the system, registers 3 objects with a total of 14 resources, subscribes to 4 resources of other (simulated) endpoints and publishes 5 resource value updates every 5 seconds. Each simulated endpoint waits a random normal-distributed time of 1-10 seconds before connecting to the CSMIM central services. This is meant to simulate the startup of a CSMIM network, with many endpoints connecting to the system at once.

The fault reporting service listens to the "csmim.obj.fault-report.1" announcements from the resource directory service, and subscribes to the respective resources. All 3 objects registered by a simulated endpoint derive from the fault reporting type and have one such resource.

The test application measures and averages the time for 3 operations:

- How long it takes for a simulated endpoint to connect to the system (authentication + MQTT connect)
- How long it takes from an endpoint requesting registration until the fault reporting service receives the announcement from the resource directory
- How long it takes from an endpoint publishing a fault report until the fault reporting service receives that report

Measurements were done on x86_64 desktop PC hardware running Linux, on 4 CPU cores with 1500 MHz. One computer executed the CSMIM central services, the other executed the test application. The computers were connected over 100 MBit switched Ethernet (cable). Here are the measured times for different numbers of simulated CSMIM endpoints:

No of simulated endpoints | Connection time | Fault report subscription time | Fault report transmission time
---|---|---|---
1 | 200ms | 50ms | 1ms
10 | 350ms | 70ms | 4ms
100 | 300ms | 200ms | 3ms
200 | 280ms | 200ms | 3ms
300 | 210ms | 400ms | 3ms
400 | 260ms | 700ms | 5ms
500 | 250ms | 750ms | 5ms
600 | 220ms | 740ms | 5ms
700 | 350ms | 800ms | 5ms
800 | 400ms | 780ms | 4ms

There is a traffic shaping component in the registration service that rejects registrations (code 503) when too many requests come in at the same time. This mechanism kicks in somewhere between 300 and 400 simulated endpoints. Rejected endpoints retry after waiting for a few seconds. You can see this in the noticeable increase in the "fault report subscription time", which includes registration. A similar mechanism for authentication seems to become effective around 700 simulated endpoints.

To sum it up, the startup phase of a CSMIM network puts considerable load on this implementation of the CSMIM central services. On the contrary, the MQTT broker core appears to handle the traffic volume quite well. There is no indication here that the proposed resource directory and fault reporting mechanism would unduly affect CSMIM network performance.

**Comment by uwe-salomon** [comment]
See the descriptions in the `types/csmim.obj.directory.1.yaml` file in the connected pull request which specify the resource directory service.

Notable difference: I removed the splitting of the version suffix, that is, the message from the above example would be published under the MQTT `topic v1/data/core/csmim/directory/csmim.obj/fault-report.1/crew/galley/left/x5`. Rationale:

- The version suffix is not mandatory for object type IDs as per ARINC 853.
- Use of a version suffix does not imply any semantic meaning, in particular it does *not* imply that version 2 is somehow compatible to version 1. From the point of view of CSMIM, an object type ID is a string.
- A client would typically subscribe to `core/csmim/directory/csmim.obj/...` for one specific version ID, namely the one it is interested in. This subscription would also catch other versions that are actually compatible, because these would be derived from the base version. Hence, a wildcard subscription for any version does not serve a purpose in many cases.
- For the same reason we would hesitate to add more versions of an object type to the resource directory's list of types of interest. This instantly doubles the amount of messages, but with little benefit.
- This way it is simpler.
